nexus/src/app/sagas/instance_update/mod.rs

```diff
@@ -0,0 +1,2514 @@
+//! the sled-agents responsible for the source and target Propolis processes,
+//! respectively. If a sled-agent pushes a state update to a particular Nexus
+//! instance and that Nexus fails to write the state to the database, that isn't
+//! the end of the world: the sled-agent can simply retry with a different
+//! Nexus, and the generation number, which is incremented exclusively by the
+//! sled-agent, ensures that state changes are idempotent and ordered. If a
+//! faulty Nexus were to return 200 OK to a sled-agent's call to
+//! `cpapi_instances_put` but choose to simply eat the received instance state
+//! update rather than writing it to the database, even *that* wouldn't
+//! necessarily mean that the state change was gone forever: the
+//! `instance-watcher` background task on another Nexus instance would
+//! eventually poll the sled-agent's state and observe any changes that were
+//! accidentally missed. This is all very neat and tidy, and we should feel
+//! proud of ourselves for having designed such a nice little system.
+//!
+//! Unfortunately, when we look beyond the `vmm` and `migration` tables, things
+//! rapidly become interesting (in the "may you live in interesting times"
+//! sense). The `instance` record *cannot* be owned exclusively by anyone. The
+//! logical instance state it represents is a gestalt that may consist of state
+//! that exists in multiple VMM processes on multiple sleds, as well as
+//! control-plane operations triggered by operator inputs and performed by
+//! multiple Nexus instances. This is, as they say, "hairy". The neat and tidy
+//! little state updates published by sled-agents to Nexus in the previous
+//! paragraph may, in some cases, represent a state transition that also
+//! requires changes to the `instance` table: for instance, a live migration may
+//! have completed, necessitating a change in the instance's `propolis_id` to
+//! point to the new VMM.
+//!
+//! Oh, and one other thing: the `instance` table record in turn logically
+//! "owns" other resources, such as the virtual-provisioning counters that
+//! represent rack-level resources allocated to the instance, and the instance's
+//! network configuration. When the instance's state changes, these resources
+//! owned by the instance may also need to be updated, such as changing the
+//! network configuration to point at an instance's new home after a successful
+//! migration, or deallocating virtual provisioning counters when an instance is
+//! destroyed. Naturally, these updates must also be performed reliably and
+//! inconsistent states must be avoided.
+//!
+//! Thus, we arrive here, at the instance-update saga.
+//!
+//! ## Theory of Operation
+//!
+//! In order to ensure that changes to the state of an instance are handled
+//! reliably, we require that all mutations of an instance` record are performed
+//! by a saga. The following sagas currently touch the `instance` record:
```

Comment:
Nit: I'm not sure this is true for *all* mutations of an instance record. There are some things in there that are user-defined properties--the cloud-init user data, virtual platform family (once that's implemented), possibly even the machine shape--that I think we can change with a single DB query that doesn't require a saga. What sets these operations is that they have several distinct steps that can't be made totally atomic, and we need a way to make sure that they either all complete or all unwind--which is, of course, just what sagas are for.

Reply:
Yeah, I think that's fair, I'll try to reword this.